project(t_t_t)

vitttoria opened the pull request:
Добрый вечер!

Возникли проблемы с полем. Не могу понять, почему оно выводится не такое, как в описании. Если я в отдельном файле проверяю создание поле, то оно рисуется, как нужно.

vitttoria committed:
project(t_t_t)

Добрый вечер!

Возникли проблемы с полем. Не могу понять, почему оно выводится не такое, как в описании.
Если я в отдельном файле проверяю создание поле, то оно рисуется, как нужно.

EgorOb commented on task_tic_tac_toe_new.py:
```diff
@@ -0,0 +1,175 @@
+        print("Есть желание повторить игру? да/нет")
+        continue_game = "да" or "нет"
+        if continue_game == "да":
+            game(table, size, player)
```
Только тут тогда нужно заново инициализировать поле, так как иначе игра будет на заполненном поле

EgorOb commented on task_tic_tac_toe_new.py:
```diff
@@ -0,0 +1,175 @@
+def begin():
+    global who_play
+    size = get_int_val("Введи размер поля\n")
+    empty_place = f'| |'  # пустые места заполняются числами от 1 до размера поля к квадрате
```
`empty_place = ' '`
так как вертикальные линии есть в отрисовке

EgorOb commented on task_tic_tac_toe_new.py:
```diff
@@ -0,0 +1,175 @@
+    :param table:
+    :return:
+    """
+    for win_combination in table:
```
Так выведутся только строки.
```
for col in zip(*table):
    ...
```
Так можно пробежаться по столбцам
Но нужно как-то пробежаться по диагоналям. Для главной диагонали условие будет равенства индекса строки и столбца [i][i]. Для побочной [i][size-1-i], где size размер поля
Можно как-то так
```
win_tuple = ("XXX", "000")
for row in table: # бежим по строкам
    if "".join(str(row)) in win_tuple: # если то, что записано на поле является или "XXX" или "000", то выиграли
        return True
for col in zip(*table): # бежим по столбцам
    if "".join(str(col)) in win_tuple: 
        return True
# ну и как-то проверить диагонали
...
return False
```

EgorOb reviewed (approved):
Нужно как-то доделать is_win, чтобы диагонали тоже рассматривались и будет всё хорошо

vitttoria committed:
Update task_tic_tac_toe_new.py

Скорректировала код с учетом замечаний. 
Есть небольшие вопросы:
1. Я понимаю, как считать выигрыш одного из игроков, если смотреть на бумаге. 
При размере поля 3х3. Выигрышные позиции по диагонали [00, 11, 22], [20, 11, 02]
Расписывая математически, мы берем и смотрим позиции
1. (строка + 1, столбец + 1), пока не достигнем  последнего столбца и строки поля
2. (строка = (размер поля - 1), столбец + 1), пока не достигнем последнего столбца поля
Затрудняюсь описать эту логику кодом.

EgorOb commented on task_tic_tac_toe_new.py:
```diff
@@ -0,0 +1,177 @@
+    for col in zip(*table):  # бежим по столбцам
+        if "".join(str(col)) in win_tuple:
+            return True
+    for row, col in table:
```
так можно заменить поиск по диагонали
```
len_t = len(table)
major_diag = []
minor_diag = []
for i in range(len_t):
    major_diag.append(table[i][i])
    minor_diag.append(table[i][len_t - 1 - i])
if "".join(str(major_diag)) in win_tuple:
    return True
if "".join(str(minor_diag)) in win_tuple:
    return True
```

EgorOb reviewed (approved):
Поправил как можно проверить диагональные элементы. Зачёл эту работу. Зачёт поставил
